src/security/PeerOptions.cc

```diff
@@ -801,6 +806,19 @@ Security::PeerOptions::updateSessionOptions(Security::SessionPointer &s)
 #endif
 }
 
+#if OPENSSL_KTLS_SUPPORT
+void
+Security::PeerOptions::updateSessionOptionsOnlyKtls(Security::SessionPointer &s)
+{
+    parseOptions();
+    
+    debugs(83, 5, "set OpenSSL options (KTLS only) for session=" << s << ", Options=" << (parsedOptions & SSL_OP_ENABLE_KTLS));
+    SSL_clear_options(s.get(), SSL_OP_ENABLE_KTLS);
```

Comment:
Why do we need to clear options before setting them in this context, but not in other contexts where SSL_set_options() is called? PR commit 124aa58b201c3d25c0acccdc31178606f1cb820c message suggests that this was done "to support squid -k reconfigure", but reconfiguration should not affect _existing_ SSL objects. I have not checked, but I would expect this `s` to be a new SSL object, created after Squid started to reconfigure itself.

Reply:
> "but I would expect this s to be a new SSL object, created after Squid started to reconfigure itself"

This was what I thought (expected) at first too, but actually, the test result was different.
I suspect some underlying issue around here (either in squid or openssl).

Reply:
Here are the logs for that.


<1> Initial condition

Config with  tls_outgoing_options options=ENABLE_KTLS and http_port options=ENABLE_KTLS

```
:22.815 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bbe6000, Options=8
:22.815 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bbe6000
:22.822 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bc5f000, Options=8
:22.822 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bc5f000
:22.831 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a29a000, Options=8
:22.831 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a29a000
:22.839 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a473000, Options=8
:22.839 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a473000
:22.847 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87c1d5000, Options=8
:22.847 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87c1d5000
:22.863 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87c1e9000, Options=8
:22.863 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87c1e9000
:22.904 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a479000, Options=8
:22.904 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a479000
:22.927 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a47d000, Options=8
:22.927 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a47d000
:23.488 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a473000, Options=8
:23.488 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a473000
:23.499 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a32c000, Options=8
:23.499 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a32c000
:23.513 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a32e000, Options=8
:23.513 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a32e000
:23.525 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a332000, Options=8
:23.525 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a332000
:23.533 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a334000, Options=8
:23.534 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a334000
:23.542 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a413000, Options=8
:23.542 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a413000
:23.552 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a47f000, Options=8
:23.552 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a47f000
:23.749 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bb1a000, Options=8
:23.749 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bb1a000
:23.753 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a198000, Options=8
:23.753 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a198000
:23.760 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a422000, Options=8
:23.760 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a422000
:23.772 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a425000, Options=8
:23.772 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a425000
:23.791 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bb74000, Options=8
:23.791 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bb74000


Options is correctly recognized as 8.
Every session is created with KTLS. (expected behavior)
```


<2> without SSL_clear_options()

remove tls_outgoing_options options=ENABLE_KTLS and http_port options=ENABLE_KTLS
squid -k reconfigure

```
:04.447 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x82ca51000, Options=0
:04.447 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x82ca51000
:04.455 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a28d000, Options=0
:04.455 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a28d000
:04.464 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bf3f000, Options=0
:04.464 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bf3f000
:04.486 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bf41000, Options=0
:04.486 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bf41000
:04.494 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a2c4000, Options=0
:04.494 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a2c4000
:04.507 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bef5000, Options=0
:04.507 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bef5000
:04.557 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a329000, Options=0
:04.557 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87a329000
:04.677 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bf41000, Options=0
:04.677 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87bf41000
:05.019 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bb71000, Options=0
:05.019 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bb71000
:05.031 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87be9e000, Options=0
:05.031 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87be9e000
:05.045 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bea0000, Options=0
:05.045 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bea0000
:05.057 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bea2000, Options=0
:05.058 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bea2000
:05.066 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bff4000, Options=0
:05.066 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87bff4000
:05.076 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87be7a000, Options=0
:05.076 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87be7a000
:05.086 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87be7c000, Options=0
:05.086 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87be7c000
:05.283 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87be7e000, Options=0
:05.283 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87be7e000
:05.287 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87bf4c000, Options=0
:05.287 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87bf4c000
:05.295 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a4e9000, Options=0
:05.295 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a4e9000
:05.314 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a4eb000, Options=0
:05.314 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a4eb000
:05.326 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87a4ed000, Options=0
:05.326 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 1 for TLS session=0x87a4ed000


Options is correctly recognized as 0.
However, some of the sessons are created with KTLS. (not expected behavior)
Perhaps the server side state remains even after the reconfiguration.
```

<3> The same test with SSL_clear_options() (Initial condition is same)

remove tls_outgoing_options options=ENABLE_KTLS and http_port options=ENABLE_KTLS
squid -k reconfigure

```
:42.426 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d364000, Options=0
:42.426 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d364000
:42.441 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d366000, Options=0
:42.441 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d366000
:42.444 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d377000, Options=0
:42.444 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d377000
:42.462 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d513000, Options=0
:42.462 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d513000
:42.467 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87e92e000, Options=0
:42.467 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87e92e000
:42.486 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87eb7f000, Options=0
:42.486 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87eb7f000
:42.652 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87ec14000, Options=0
:42.652 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87ec14000
:42.973 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d366000, Options=0
:42.973 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d366000
:42.980 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d2b7000, Options=0
:42.980 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d2b7000
:43.124 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d2e2000, Options=0
:43.124 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d2e2000
:43.128 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d3ad000, Options=0
:43.128 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d3ad000
:43.452 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87ec14000, Options=0
:43.453 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87ec14000
:43.551 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d418000, Options=0
:43.551 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d418000
:43.567 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87ea83000, Options=0
:43.567 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87ea83000
:43.571 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d41a000, Options=0
:43.571 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d41a000
:43.581 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87eb7f000, Options=0
:43.581 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87eb7f000
:43.589 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87e99e000, Options=0
:43.589 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87e99e000
:43.591 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87ea76000, Options=0
:43.591 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87ea76000
:43.613 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d3a8000, Options=0
:43.613 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d3a8000
:43.728 kid1| 83,5| PeerOptions.cc(815) updateSessionOptionsOnlyKtls: set OpenSSL options (KTLS only) for session=0x87d3aa000, Options=0
:43.728 kid1| 83,5| Session.cc(142) CreateSession: KTLS: 0 for TLS session=0x87d3aa000


Options is correctly recognized as 0.
Every session is created without KTLS. (expected behavior)
```